backend/tests/end_to_end/infrastructures/apps/load/test_admin_panel.py

```diff
@@ -0,0 +1,231 @@
+import csv
+import io
+from http import HTTPStatus
+
+from django.db import transaction
+from django.http import HttpRequest
+from freezegun import freeze_time
+
+from infrastructures.apps.load import admin
+from infrastructures.apps.load.models import Data
+from tests import model_factories
+
+from ....dtos import APIClientData, Client
+from ....utils import get_url
+
+
+def test_saving_new_data(
+    unauthenticated_client: APIClientData,
+    data_admin_panel: admin.DataAdmin,
+):
+    # Given
+    client = unauthenticated_client.client
+
+    data: Data = model_factories.DataFactory.create()
+
+    form = admin.DataAdminForm(
+        data={
+            "data": {
+                "full_name": data.data["full_name"],
+                "age": data.data["age"],
+                "is_satisfied": data.data["is_satisfied"],
+            },
+        }
+    )
+
+    with transaction.atomic():
+        form.is_valid()
+        assert not form.errors
+    change = False
+
+    # When
+    data_admin_panel.save_model(
+        request=HttpRequest(), obj=data, form=form, change=change
+    )
+
+    # Then
+    response = client.get(
+        get_url(
+            path_name="load-detail",
+            path_params={"pk": data.id},
+        )
+    )
+    assert response.status_code == HTTPStatus.OK
+    assert response.data["full_name"] == data.data["full_name"]
+
+
+def test_editing_data(
+    data_admin_panel: admin.DataAdmin,
+):
+    # Given
+    data: Data = model_factories.DataFactory.create()
+
+    new_full_name = "New Full_Name"
+    form = admin.DataAdminForm(
+        data={
+            "data": {
+                "full_name": new_full_name,
+                "age": data.data["age"],
+                "is_satisfied": data.data["is_satisfied"],
+            },
+        },
+        instance=data,
+    )
+
+    with transaction.atomic():
+        form.is_valid()
+        assert not form.errors
+    change = True
+
+    # When
+    data_admin_panel.save_model(
+        request=HttpRequest(), obj=data, form=form, change=change
+    )
+
+    # Then
+    data.refresh_from_db()
+    assert data.data["full_name"] == new_full_name
+
+
+def test_delete_data(
+    unauthenticated_client: APIClientData,
+    data_admin_panel: admin.DataAdmin,
+):
+    # Given
+    client = unauthenticated_client.client
+
+    data: Data = model_factories.DataFactory.create()
+
+    # When
+    data_admin_panel.delete_model(
+        request=HttpRequest(),
+        obj=data,
+    )
+
+    # Then
+    response = client.get(
+        get_url(
+            path_name="load-detail",
+            path_params={"pk": data.id},
+        )
+    )
+    assert response.status_code == HTTPStatus.NOT_FOUND
+
+
+def test_admin_form_raises_validation_error_when_data_is_not_dict():
```

Comment:
Is this test needed? Or am I missing something?

Reply:
well, I think I can delete it, as dict is enforced.